### PR TITLE
Allow custom file extensions for EmitOptions

### PR DESCRIPTION
**Pull request by steven-johnson**
Some build environments need to use nonstandard extensions for some Generator products; add an option to EmitOptions to allow remapping the default file extensions for build products. (This can be worked around by using ‘rename’ after the fact, but is awkward.)

### CODE REVIEW

**Comment by steven-johnson** [comment]
(Also did a drive-by addition of "cpp" as a valid option to -e)

**Comment by steven-johnson** [comment]
Travis build appears to be an apparently-unrelated failure?

"The command "sudo apt-get -y install gcc-4.8 g++-4.8 libedit-dev" failed and exited with 100 during ."

File: src/Generator.cpp [hunk 1]
```diff
@@ -47,15 +47,17 @@ const std::map<std::string, Halide::Type> &get_halide_type_enum_map() {
 }
 
 int generate_filter_main(int argc, char **argv, std::ostream &cerr) {
-    const char kUsage[] = "gengen [-g GENERATOR_NAME] [-f FUNCTION_NAME] [-o OUTPUT_DIR] [-r RUNTIME_NAME] [-e EMIT_OPTIONS] "
+    const char kUsage[] = "gengen [-g GENERATOR_NAME] [-f FUNCTION_NAME] [-o OUTPUT_DIR] [-r RUNTIME_NAME] [-e EMIT_OPTIONS] [-x EXTENSION_OPTIONS] "
                           "target=target-string [generator_arg=value [...]]\n\n"
                           "  -e  A comma separated list of optional files to emit. Accepted values are "
-                          "[assembly, bitcode, stmt, html]\n";
+                          "[assembly, bitcode, stmt, html, cpp]\n"
+                          "  -x  A comma separated list of extension=extension pairs to subsitute during file naming.\n";
```
Comment (dsharletg) [inline src/Generator.cpp]: What does 'extension=extension' mean?

Comment (steven-johnson) [inline src/Generator.cpp]: It means "I neglected to go back and clean this up before sending the PR". Oops. Will improve.

**Comment by dsharletg** [comment]
LGTM aside from the help message on the extension mapping, not sure I would be able to use that to figure out how to change the extension without some trial and error.

**Pull request merged**